src/Analyser/NodeScopeResolver.php

```diff
@@ -1843,74 +1841,60 @@ function (MutatingScope $scope) use ($expr, $nodeCallback, $context): Expression
 				&& in_array($functionReflection->getName(), ['array_push', 'array_unshift'], true)
 				&& count($expr->getArgs()) >= 2
 			) {
-				$argumentTypes = [];
-				foreach (array_slice($expr->getArgs(), 1) as $callArg) {
-					$callArgType = $scope->getType($callArg->value);
-					if ($callArg->unpack) {
-						$iterableValueType = $callArgType->getIterableValueType();
-						if ($iterableValueType instanceof UnionType) {
-							foreach ($iterableValueType->getTypes() as $innerType) {
-								$argumentTypes[] = $innerType;
+				$arrayArg = $expr->getArgs()[0]->value;
+				$arrayType = $scope->getType($arrayArg);
+				$callArgs = array_slice($expr->getArgs(), 1);
+				$prepend = $functionReflection->getName() === 'array_unshift';
+
+				if ($arrayType instanceof ConstantArrayType) {
+					$arrayTypeBuilder = $prepend ? ConstantArrayTypeBuilder::createEmpty() : ConstantArrayTypeBuilder::createFromConstantArray($arrayType);
+					foreach ($callArgs as $callArg) {
+						$callArgType = $scope->getType($callArg->value);
+						if ($callArg->unpack) {
+							if ($callArgType->isIterableAtLeastOnce()->no()) {
+								continue;
```

Comment:
I think this is only needed for empty arrays with nevertype. Weird that this is not existing in the other block. Will check later